### PR TITLE
add validation to prevent namespace to move if used by an app

### PR DESCRIPTION
**Pull request by StrongMonkey**

### CODE REVIEW

**Comment by StrongMonkey** [comment]
https://github.com/rancher/rancher/issues/14845

**Comment by StrongMonkey** [comment]
https://github.com/rancher/rancher/issues/14845

**Comment by StrongMonkey** [comment]
requires https://github.com/rancher/types/pull/562

**Comment by StrongMonkey** [comment]
This pr also add a separate action called `move` on namespace and make projectID non-updatable. The corresponding UI part also needs to be changed.

**Comment by StrongMonkey** [comment]
@cjellick this PR has following changes:
1. if namespace is used by an app, annotation `cattle.io/appIds` is added with the appID.
2. If namespace has that annotation, the move action will be removed.
3. If app is removed, that annotation also gets removed.

File: pkg/api/customization/namespace/namespace.go [hunk 1]
```diff
@@ -44,3 +52,46 @@ func ProjectMap(apiContext *types.APIContext, refresh bool) (map[string]string,
...
+		userContext, err := w.ClusterManager.UserContext(clusterID)
+		if err != nil {
+			return err
+		}
+		nsClient := userContext.Core.Namespaces("")
+		ns, err := nsClient.Get(apiContext.ID, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		ns.Annotations[projectIDFieldLabel] = convert.ToString(actionInput["projectId"])
```
Comment (dramich) [inline pkg/api/customization/namespace/namespace.go]: Add a safety check, if the annotation exists we shouldn't be moving the namespace

File: pkg/controllers/user/helm/controller.go [hunk 2]
```diff
@@ -153,6 +175,26 @@ func (l *Lifecycle) Remove(obj *v3.App) (*v3.App, error) {
...
+	appIds := strings.Split(ns.Annotations[AppIDsLabel], ",")
+	appAnno := ""
+	for _, appID := range appIds {
+		if appID == obj.Name {
+			continue
+		}
+		appAnno += appID + ","
+	}
+	appAnno = strings.TrimSuffix(appAnno, ",")
+	ns.Annotations[AppIDsLabel] = appAnno
```
Comment (dramich) [inline pkg/controllers/user/helm/controller.go]: You could add this to an else block

File: pkg/controllers/user/helm/controller.go [hunk 3]
```diff
@@ -102,7 +108,23 @@ func (l *Lifecycle) Updated(obj *v3.App) (*v3.App, error) {
...
+		return result, err
+	}
+	ns, err := l.NsClient.Get(obj.Spec.TargetNamespace, metav1.GetOptions{})
+	if err != nil {
+		return result, err
+	}
+	if ns.Annotations[AppIDsLabel] == "" {
+		ns.Annotations[AppIDsLabel] = obj.Name
+	} else {
+		ns.Annotations[AppIDsLabel] = ns.Annotations[AppIDsLabel] + "," + obj.Name
```
Comment (dramich) [inline pkg/controllers/user/helm/controller.go]: Add a check to see if the appID already exists before appending

**Comment by cjellick** [comment]
LGTM

**Pull request merged**